src/mbgl/annotation/annotation_manager.cpp

```diff
@@ -92,21 +100,27 @@ Update AnnotationManager::update(const AnnotationID& id, const SymbolAnnotation&
 }
 
 Update AnnotationManager::update(const AnnotationID& id, const LineAnnotation& annotation, const uint8_t maxZoom) {
+    mutex.lock();
     auto it = shapeAnnotations.find(id);
     if (it == shapeAnnotations.end()) {
         assert(false); // Attempt to update a non-existent shape annotation
         return Update::Nothing;
     }
+    mutex.unlock();
+
     removeAndAdd(id, annotation, maxZoom);
```

Comment:
There should be a lock in `removeAndAdd` around both the remove and add, not merely a lock in `add` and a lock in `remove`. Otherwise, a tile could get updated in between the remove and the add, producing a flickering annotation.